extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesTaskRunner.java

```diff
@@ -134,46 +134,37 @@ public Optional<InputStream> streamTaskLog(String taskid, long offset)
   @Override
   public ListenableFuture<TaskStatus> run(Task task)
   {
-    return tasks.computeIfAbsent(
-        task.getId(), k -> new KubernetesWorkItem(task, exec.submit(() -> runTask(task)))
-    ).getResult();
+    return runOrJoinTask(task, true);
   }
 
-  protected ListenableFuture<TaskStatus> joinAsync(Task task)
+  protected ListenableFuture<TaskStatus> runOrJoinTask(Task task, boolean run)
```

Comment:
It was better to have this as two separate methods, seemed more readable and easy to understand.

Reply:
yeah agree, we can just put the synchronized block in (revert parts of that previous pr i made to remove them) without changing any function names

Reply:
Done. My comment as above:

> My thinking was since we added synchronized block (or lock in the previous commit), it's better to have a single place/method to have this complexity. But agree this may make the code less readable.